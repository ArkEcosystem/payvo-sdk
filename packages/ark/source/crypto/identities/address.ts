import { Base58Check, Hash } from "@payvo/sdk-cryptography";

import { Network } from "../interfaces/networks";
import { MultiSignatureAsset, KeyPair } from "./contracts";
import { PublicKeyError } from "./errors";
import { getPubKeyHash } from "./helpers.js";
import { PublicKey } from "./public-key";

export class Address {
	public static fromPassphrase(passphrase: string, network?: Network): string {
		return Address.fromPublicKey(PublicKey.fromPassphrase(passphrase), network);
	}

	public static fromPublicKey(publicKey: string, network?: Network): string {
		if (!PublicKey.verify(publicKey)) {
			throw new PublicKeyError(publicKey);
		}

		const buffer: Buffer = Hash.ripemd160(Buffer.from(publicKey, "hex"));
		const payload: Buffer = Buffer.alloc(21);

		payload.writeUInt8(getPubKeyHash(network), 0);
		buffer.copy(payload, 1);

		return this.fromBuffer(payload);
	}

	public static fromWIF(wif: string, network?: Network): string {
		return Address.fromPublicKey(PublicKey.fromWIF(wif, network), network);
	}

	public static fromMultiSignatureAsset(asset: MultiSignatureAsset, network?: Network): string {
		return this.fromPublicKey(PublicKey.fromMultiSignatureAsset(asset), network);
	}

	public static fromPrivateKey(privateKey: KeyPair, network?: Network): string {
		return Address.fromPublicKey(privateKey.publicKey, network);
	}

	public static fromBuffer(buffer: Buffer): string {
		return Base58Check.encode(buffer);
	}

	public static toBuffer(
		address: string,
		network?: Network,
	): { addressBuffer: Buffer; addressError?: string } {
		const buffer: Buffer = Base58Check.decode(address);
		const result: { addressBuffer: Buffer; addressError?: string } = {
			addressBuffer: buffer,
		};

		if (buffer[0] !== getPubKeyHash(network)) {
			result.addressError = `Expected address network byte ${getPubKeyHash(network)}, but got ${buffer[0]}.`;
		}

		return result;
	}

	public static validate(address: string): boolean {
		try {
			return Base58Check.decode(address)[0] === getPubKeyHash();
		} catch (err) {
			return false;
		}
	}
}
