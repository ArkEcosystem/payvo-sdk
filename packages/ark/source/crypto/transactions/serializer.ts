import { ByteBuffer } from "../crypto/buffer.js";

import { TransactionTypeGroup } from "../enums.js";
import { TransactionVersionError } from "../errors.js";
import { ISerializeOptions, ITransaction, ITransactionData } from "../interfaces/index.js";
import { configManager } from "../managers/config.js";
import { isSupportedTransactionVersion } from "../utils.js";
import { TransactionTypeFactory } from "./types/factory.js";

// Reference: https://github.com/ArkEcosystem/AIPs/blob/master/AIPS/aip-11.md
export class Serializer {
	public static getBytes(transaction: ITransactionData, options: ISerializeOptions = {}): Buffer {
		const version: number = transaction.version || 2;

		if (options.acceptLegacyVersion || options.disableVersionCheck || isSupportedTransactionVersion(version)) {
			return this.serialize(TransactionTypeFactory.create(transaction), options);
		} else {
			throw new TransactionVersionError(version);
		}
	}

	/**
	 * Serializes the given transaction according to AIP11.
	 */
	public static serialize(transaction: ITransaction, options: ISerializeOptions = {}): Buffer {
		const buffer: ByteBuffer = new ByteBuffer(512);

		this.serializeCommon(transaction.data, buffer);
		this.serializeVendorField(transaction, buffer);

		const serialized: ByteBuffer | undefined = transaction.serialize(options);

		if (!serialized) {
			throw new Error();
		}

		const typeBuffer: ByteBuffer = serialized.flip();
		buffer.append(typeBuffer.toBuffer());

		this.serializeSignatures(transaction.data, buffer, options);

		const flippedBuffer: Buffer = buffer.flip().toBuffer();
		transaction.serialized = flippedBuffer;

		return flippedBuffer;
	}

	private static serializeCommon(transaction: ITransactionData, buffer: ByteBuffer): void {
		transaction.version = transaction.version || 0x01;
		if (transaction.typeGroup === undefined) {
			transaction.typeGroup = TransactionTypeGroup.Core;
		}

		console.log(buffer.toBuffer().byteLength);

		buffer.writeByte(0xff);
		buffer.writeByte(transaction.version);
		buffer.writeByte(transaction.network || configManager.get("network.pubKeyHash"));
		buffer.writeUint32(transaction.typeGroup);
		buffer.writeUint16(transaction.type);

		if (transaction.nonce) {
			buffer.writeUint64(transaction.nonce.toString());
		}

		if (transaction.senderPublicKey) {
			buffer.append(transaction.senderPublicKey, "hex");
		}

		buffer.writeUint64(transaction.fee.toString());
	}

	private static serializeVendorField(transaction: ITransaction, buffer: ByteBuffer): void {
		if (transaction.hasVendorField()) {
			const { data }: ITransaction = transaction;

			if (data.vendorField) {
				const vf: Buffer = Buffer.from(data.vendorField, "utf8");
				buffer.writeByte(vf.length);
				buffer.append(vf);
			} else {
				buffer.writeByte(0x00);
			}
		} else {
			buffer.writeByte(0x00);
		}
	}

	private static serializeSignatures(
		transaction: ITransactionData,
		buffer: ByteBuffer,
		options: ISerializeOptions = {},
	): void {
		if (transaction.signature && !options.excludeSignature) {
			buffer.append(transaction.signature, "hex");
		}

		const secondSignature: string | undefined = transaction.secondSignature || transaction.signSignature;

		if (secondSignature && !options.excludeSecondSignature) {
			buffer.append(secondSignature, "hex");
		}

		if (transaction.signatures && !options.excludeMultiSignature) {
			buffer.append(transaction.signatures.join(""), "hex");
		}
	}
}
