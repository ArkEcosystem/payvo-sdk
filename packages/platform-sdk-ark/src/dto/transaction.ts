import { Contracts, DTO } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.id;
	}

	public blockId(): string | undefined {
		return this.data.blockId;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.fromUnix(this.data.timestamp.unix);
	}

	public confirmations(): BigNumber {
		return BigNumber.make(this.data.confirmations);
	}

	public sender(): string {
		return this.data.sender;
	}

	public recipient(): string {
		return this.data.recipient;
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		if (!this.isMultiPayment()) {
			return [];
		}

		return this.data.asset.payments.map((payment: { recipientId: string; amount: string }) => ({
			address: payment.recipientId,
			amount: BigNumber.make(payment.amount),
		}));
	}

	public amount(): BigNumber {
		if (this.isMultiPayment()) {
			return this.data.asset.payments.reduce(
				(sum: BigNumber, { amount }: { amount: string }) => sum.plus(amount),
				BigNumber.ZERO,
			);
		}

		return BigNumber.make(this.data.amount);
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.fee);
	}

	public memo(): string | undefined {
		return this.data.vendorField;
	}

	public asset(): Record<string, unknown> {
		return this.data.asset;
	}

	public isConfirmed(): boolean {
		return this.confirmations().isGreaterThanOrEqualTo(51);
	}

	public isSent(): boolean {
		return [this.getMeta("address"), this.getMeta("publicKey")].includes(this.sender());
	}

	public isReceived(): boolean {
		return [this.getMeta("address"), this.getMeta("publicKey")].includes(this.recipient());
	}

	public isTransfer(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 0;
	}

	public isSecondSignature(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 1;
	}

	public isDelegateRegistration(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 2;
	}

	public isVote(): boolean {
		const isVote = this.data.typeGroup === 1 && this.data.type === 3;

		if (!isVote) {
			return false;
		}

		return (this.asset().votes as string[])[0].startsWith("+");
	}

	public isUnvote(): boolean {
		const isVote = this.data.typeGroup === 1 && this.data.type === 3;

		if (!isVote) {
			return false;
		}

		return (this.asset().votes as string[])[0].startsWith("-");
	}

	public isMultiSignature(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 4;
	}

	public isIpfs(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 5;
	}

	public isMultiPayment(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 6;
	}

	public isDelegateResignation(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 7;
	}

	public isHtlcLock(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 8;
	}

	public isHtlcClaim(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 9;
	}

	public isHtlcRefund(): boolean {
		return this.data.typeGroup === 1 && this.data.type === 10;
	}

	public isBusinessRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 0;
	}

	public isBusinessResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 1;
	}

	public isBusinessUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 2;
	}

	public isBridgechainRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 3;
	}

	public isBridgechainResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 4;
	}

	public isBridgechainUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 5;
	}

	public isEntityRegistration(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 0;
	}

	public isEntityResignation(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 1;
	}

	public isEntityUpdate(): boolean {
		return this.data.typeGroup === 2 && this.data.type === 6 && this.data.asset.action === 2;
	}
}
