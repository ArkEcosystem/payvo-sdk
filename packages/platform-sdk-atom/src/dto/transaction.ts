import { Contracts, DTO, Exceptions } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.txhash;
	}

	public blockId(): string | undefined {
		return undefined;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.make(this.data.timestamp);
	}

	public confirmations(): BigNumber {
		return BigNumber.ZERO;
	}

	public sender(): string {
		const event = this.data.events.find((event) => event.type === "message");
		const attribute = event.attributes.find((attribute) => attribute.key === "sender");

		return attribute.value;
	}

	public recipient(): string {
		const event = this.data.events.find((event) => event.type === "transfer");
		const attribute = event.attributes.find((attribute) => attribute.key === "recipient");

		return attribute.value;
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		return [];
	}

	// @ts-ignore
	public amount(): BigNumber {
		const event = this.data.events.find((event) => event.type === "transfer");
		const attribute = event.attributes.find((attribute) => attribute.key === "amount");

		return BigNumber.make(attribute.value.replace(/\D/g, ""));
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.gas_used);
	}

	public memo(): string | undefined {
		return this.data.tx.value.memo;
	}

	public asset(): Record<string, unknown> {
		return {};
	}

	public isConfirmed(): boolean {
		return false;
	}

	public isSent(): boolean {
		return false;
	}

	public isReceived(): boolean {
		return false;
	}

	public isTransfer(): boolean {
		return false;
	}

	public isSecondSignature(): boolean {
		return false;
	}

	public isDelegateRegistration(): boolean {
		return false;
	}

	public isVote(): boolean {
		return false;
	}

	public isUnvote(): boolean {
		return false;
	}

	public isMultiSignature(): boolean {
		return false;
	}

	public isIpfs(): boolean {
		return false;
	}

	public isMultiPayment(): boolean {
		return false;
	}

	public isDelegateResignation(): boolean {
		return false;
	}

	public isHtlcLock(): boolean {
		return false;
	}

	public isHtlcClaim(): boolean {
		return false;
	}

	public isHtlcRefund(): boolean {
		return false;
	}

	public isEntityRegistration(): boolean {
		return false;
	}

	public isEntityResignation(): boolean {
		return false;
	}

	public isEntityUpdate(): boolean {
		return false;
	}

	public isBusinessEntityRegistration(): boolean {
		return false;
	}

	public isBusinessEntityResignation(): boolean {
		return false;
	}

	public isBusinessEntityUpdate(): boolean {
		return false;
	}

	public isDeveloperEntityRegistration(): boolean {
		return false;
	}

	public isDeveloperEntityResignation(): boolean {
		return false;
	}

	public isDeveloperEntityUpdate(): boolean {
		return false;
	}

	public isCorePluginEntityRegistration(): boolean {
		return false;
	}

	public isCorePluginEntityResignation(): boolean {
		return false;
	}

	public isCorePluginEntityUpdate(): boolean {
		return false;
	}

	public isDesktopPluginEntityRegistration(): boolean {
		return false;
	}

	public isDesktopPluginEntityResignation(): boolean {
		return false;
	}

	public isDesktopPluginEntityUpdate(): boolean {
		return false;
	}

	public isDelegateEntityRegistration(): boolean {
		return false;
	}

	public isDelegateEntityResignation(): boolean {
		return false;
	}

	public isDelegateEntityUpdate(): boolean {
		return false;
	}

	public isLegacyBusinessRegistration(): boolean {
		return false;
	}

	public isLegacyBusinessResignation(): boolean {
		return false;
	}

	public isLegacyBusinessUpdate(): boolean {
		return false;
	}

	public isLegacyBridgechainRegistration(): boolean {
		return false;
	}

	public isLegacyBridgechainResignation(): boolean {
		return false;
	}

	public isLegacyBridgechainUpdate(): boolean {
		return false;
	}
}
