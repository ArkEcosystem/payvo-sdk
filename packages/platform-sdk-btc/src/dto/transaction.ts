import { Contracts, DTO, Exceptions } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	public id(): string {
		return this.data.txid;
	}

	public blockId(): string | undefined {
		return undefined;
	}

	public timestamp(): DateTime | undefined {
		return DateTime.make(this.data.blockTime);
	}

	public confirmations(): BigNumber {
		return BigNumber.make(this.data.confirmations);
	}

	public sender(): string {
		throw new Exceptions.NotImplemented(this.constructor.name, "sender");
	}

	public recipient(): string {
		throw new Exceptions.NotImplemented(this.constructor.name, "recipient");
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		return [];
	}

	public amount(): BigNumber {
		return BigNumber.make(this.data.value);
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.fee);
	}

	public memo(): string | undefined {
		return undefined;
	}

	public asset(): Record<string, unknown> {
		return {};
	}

	public isConfirmed(): boolean {
		return false;
	}

	public isSent(): boolean {
		return false;
	}

	public isReceived(): boolean {
		return false;
	}

	public isTransfer(): boolean {
		return false;
	}

	public isSecondSignature(): boolean {
		return false;
	}

	public isDelegateRegistration(): boolean {
		return false;
	}

	public isVote(): boolean {
		return false;
	}

	public isUnvote(): boolean {
		return false;
	}

	public isMultiSignature(): boolean {
		return false;
	}

	public isIpfs(): boolean {
		return false;
	}

	public isMultiPayment(): boolean {
		return false;
	}

	public isDelegateResignation(): boolean {
		return false;
	}

	public isHtlcLock(): boolean {
		return false;
	}

	public isHtlcClaim(): boolean {
		return false;
	}

	public isHtlcRefund(): boolean {
		return false;
	}

	public isEntityRegistration(): boolean {
		return false;
	}

	public isEntityResignation(): boolean {
		return false;
	}

	public isEntityUpdate(): boolean {
		return false;
	}

	public isBusinessEntityRegistration(): boolean {
		return false;
	}

	public isBusinessEntityResignation(): boolean {
		return false;
	}

	public isBusinessEntityUpdate(): boolean {
		return false;
	}

	public isDeveloperEntityRegistration(): boolean {
		return false;
	}

	public isDeveloperEntityResignation(): boolean {
		return false;
	}

	public isDeveloperEntityUpdate(): boolean {
		return false;
	}

	public isCorePluginEntityRegistration(): boolean {
		return false;
	}

	public isCorePluginEntityResignation(): boolean {
		return false;
	}

	public isCorePluginEntityUpdate(): boolean {
		return false;
	}

	public isDesktopPluginEntityRegistration(): boolean {
		return false;
	}

	public isDesktopPluginEntityResignation(): boolean {
		return false;
	}

	public isDesktopPluginEntityUpdate(): boolean {
		return false;
	}

	public isDelegateEntityRegistration(): boolean {
		return false;
	}

	public isDelegateEntityResignation(): boolean {
		return false;
	}

	public isDelegateEntityUpdate(): boolean {
		return false;
	}

	public isLegacyBusinessRegistration(): boolean {
		return false;
	}

	public isLegacyBusinessResignation(): boolean {
		return false;
	}

	public isLegacyBusinessUpdate(): boolean {
		return false;
	}

	public isLegacyBridgechainRegistration(): boolean {
		return false;
	}

	public isLegacyBridgechainResignation(): boolean {
		return false;
	}

	public isLegacyBridgechainUpdate(): boolean {
		return false;
	}
}
