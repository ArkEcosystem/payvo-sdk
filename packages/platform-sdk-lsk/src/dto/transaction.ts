import { Contracts, DTO } from "@arkecosystem/platform-sdk";
import { DateTime } from "@arkecosystem/platform-sdk-intl";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

export class TransactionData extends DTO.AbstractTransactionData implements Contracts.TransactionData {
	// readonly #types = {
	// 	0: "transfer",
	// 	1: "secondSignature",
	// 	2: "delegateRegistration",
	// 	3: "vote",
	// 	4: "multiSignature",
	// };

	// // TODO: remove these once live/beta handling is properly implemented
	// readonly #betanetTypes = {
	// 	8: "transfer",
	// 	9: "secondSignature",
	// 	10: "delegateRegistration",
	// 	11: "vote",
	// 	12: "multiSignature",
	// };

	public id(): string {
		return this.data.id;
	}

	public blockId(): string | undefined {
		return this.data.blockId;
	}

	public timestamp(): DateTime | undefined {
		// TODO: use a genesis timestamp that matches the network
		return DateTime.make("2016-05-24T17:00:00.000Z").addSeconds(this.data.timestamp);
	}

	public confirmations(): BigNumber {
		return BigNumber.make(this.data.confirmations);
	}

	public sender(): string {
		return this.data.senderPublicKey;
	}

	public recipient(): string {
		return this.data.recipientId;
	}

	public recipients(): Contracts.MultiPaymentRecipient[] {
		return [];
	}

	public amount(): BigNumber {
		return BigNumber.make(this.data.amount);
	}

	public fee(): BigNumber {
		return BigNumber.make(this.data.fee);
	}

	public memo(): string | undefined {
		return this.data.asset.data;
	}

	public asset(): Record<string, unknown> {
		return {};
	}

	public isConfirmed(): boolean {
		return false;
	}

	public isSent(): boolean {
		return false;
	}

	public isReceived(): boolean {
		return false;
	}

	public isTransfer(): boolean {
		return this.data.type === 8;
	}

	public isSecondSignature(): boolean {
		return this.data.type === 9;
	}

	public isDelegateRegistration(): boolean {
		return this.data.type === 10;
	}

	public isVote(): boolean {
		if (this.data.type !== 11) {
			return false;
		}

		for (const vote of this.asset().votes as string[]) {
			if (vote.startsWith("+")) {
				return true;
			}
		}

		return false;
	}

	public isUnvote(): boolean {
		if (this.data.type !== 11) {
			return false;
		}

		for (const vote of this.asset().votes as string[]) {
			if (vote.startsWith("-")) {
				return true;
			}
		}

		return false;
	}

	public isMultiSignature(): boolean {
		return this.data.type === 12;
	}

	public isIpfs(): boolean {
		return false;
	}

	public isMultiPayment(): boolean {
		return false;
	}

	public isDelegateResignation(): boolean {
		return false;
	}

	public isHtlcLock(): boolean {
		return false;
	}

	public isHtlcClaim(): boolean {
		return false;
	}

	public isHtlcRefund(): boolean {
		return false;
	}

	public isBusinessRegistration(): boolean {
		return false;
	}

	public isBusinessResignation(): boolean {
		return false;
	}

	public isBusinessUpdate(): boolean {
		return false;
	}

	public isBridgechainRegistration(): boolean {
		return false;
	}

	public isBridgechainResignation(): boolean {
		return false;
	}

	public isBridgechainUpdate(): boolean {
		return false;
	}

	public isEntityRegistration(): boolean {
		return false;
	}

	public isEntityResignation(): boolean {
		return false;
	}

	public isEntityUpdate(): boolean {
		return false;
	}
}
