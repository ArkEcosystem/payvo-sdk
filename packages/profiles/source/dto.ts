export * from "./signed-transaction.dto.js";
export * from "./transaction.collection";
export * from "./transaction.dto.js";
export * from "./transaction.mapper";
