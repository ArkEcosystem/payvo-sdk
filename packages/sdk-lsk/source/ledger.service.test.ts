import "jest-extended";

import { IoC, Services } from "@payvo/sdk";
import { openTransportReplayer, RecordStore } from "@ledgerhq/hw-transport-mocker";
import nock from "nock";

import { ledger } from "../test/fixtures/ledger";
import { createService, require } from "../test/mocking";
import { AddressService } from "./address.service";
import { ClientService } from "./client.service";
import { LedgerService } from "./ledger.service";
import { BindingType } from "./coin.contract";
import { AssetSerializer } from "./asset.serializer";
import { TransactionSerializer } from "./transaction.serializer";
import { SignedTransactionData } from "./signed-transaction.dto";
import { ConfirmedTransactionData } from "./confirmed-transaction.dto";
import { WalletData } from "./wallet.dto";

const createMockService = async (record: string) => {
	const transport = await createService(LedgerService, "lsk.mainnet", (container) => {
		container.constant(IoC.BindingType.Container, container);
		container.singleton(IoC.BindingType.AddressService, AddressService);
		container.singleton(IoC.BindingType.ClientService, ClientService);
		container.constant(IoC.BindingType.DataTransferObjects, {
			SignedTransactionData,
			ConfirmedTransactionData,
			WalletData,
		});
		container.singleton(IoC.BindingType.DataTransferObjectService, Services.AbstractDataTransferObjectService);
		container.singleton(BindingType.AssetSerializer, AssetSerializer);
		container.singleton(BindingType.TransactionSerializer, TransactionSerializer);
	});

	await transport.connect(await openTransportReplayer(RecordStore.fromString(record)));

	return transport;
};

describe("connect", () => {
	it("should throw error with unexpected input", async () => {
		const transport = await createService(LedgerService, "lsk.mainnet", (container) => {
			container.constant(IoC.BindingType.Container, container);
			container.singleton(IoC.BindingType.AddressService, AddressService);
			container.singleton(IoC.BindingType.ClientService, ClientService);
			container.constant(IoC.BindingType.DataTransferObjects, {
				SignedTransactionData,
				ConfirmedTransactionData,
				WalletData,
			});
			container.singleton(IoC.BindingType.DataTransferObjectService, Services.AbstractDataTransferObjectService);
			container.singleton(BindingType.AssetSerializer, AssetSerializer);
			container.singleton(BindingType.TransactionSerializer, TransactionSerializer);
		});

		await expect(() =>
			transport.connect({
				open: () => {
					throw new Error("cannot open");
				},
			}),
		).rejects.toThrow();
	});
});

describe("disconnect", () => {
	it("should pass with a resolved transport closure", async () => {
		const lsk = await createMockService("");

		await expect(lsk.disconnect()).resolves.toBeUndefined();
	});
});

describe("getVersion", () => {
	it("should pass with an app version", async () => {
		const lsk = await createMockService(ledger.appVersion.record);

		await expect(lsk.getVersion()).resolves.toEqual(ledger.appVersion.result);
	});
});

describe("getPublicKey", () => {
	it("should pass with a compressed publicKey", async () => {
		const lsk = await createMockService(ledger.publicKey.record);

		await expect(lsk.getPublicKey(ledger.bip44.path)).resolves.toEqual(ledger.publicKey.result);
	});
});

describe("signTransaction", () => {
	it("should pass with a signature", async () => {
		const lsk = await createMockService(ledger.transaction.record);

		await expect(
			lsk.signTransaction(ledger.bip44.path, Buffer.from(ledger.transaction.payload, "hex")),
		).resolves.toEqual(ledger.transaction.result);
	});
});

describe("signMessage", () => {
	it("should pass with a signature", async () => {
		const lsk = await createMockService(ledger.message.record);

		await expect(lsk.signMessage(ledger.bip44.path, Buffer.from(ledger.message.payload, "hex"))).resolves.toEqual(
			ledger.message.result,
		);
	});
});

describe("scan", () => {
	afterEach(() => nock.cleanAll());

	beforeAll(() => nock.disableNetConnect());

	it("should return scanned wallet", async () => {
		nock(/.+/)
			.get("/api/v2/accounts")
			.query({ address: "lsk8s6v2pdnxvab9oc42wbhvtb569jqg2ubjxgvvj" })
			.reply(200, await require("../test/fixtures/client/wallet-0.json"))
			.get("/api/v2/accounts")
			.query({ address: "lskbh47p4ts33c6c5pjvwa32424qr8px8pwfx8e4s" })
			.reply(200, await require("../test/fixtures/client/wallet-1.json"))
			.get("/api/v2/accounts")
			.query({ address: "lskksmfa2q2evtwmfneaon79u9hv7a3saokuy9tv9" })
			.reply(200, await require("../test/fixtures/client/wallet-2.json"));

		const lsk = await createMockService(ledger.wallets.record);

		const walletData = await lsk.scan();

		expect(Object.keys(walletData)).toHaveLength(4); // 3 + 1 cold wallet
		expect(walletData).toMatchSnapshot();
	});

	it("should allow to pass a startPath", async () => {
		const lsk = await createMockService(ledger.wallets.record2);

		const walletData = await lsk.scan({ startPath: "44'/134'/10'/0/0" });

		expect(Object.keys(walletData)).toHaveLength(1);
		expect(walletData).toMatchSnapshot();
	});

	it("should support legacy", async () => {
		const lsk = await createMockService(ledger.wallets.record);

		const walletData = await lsk.scan({ useLegacy: true });

		expect(Object.keys(walletData)).toHaveLength(1);
		expect((Object.values(walletData) as any[])[0].data.address).not.toStartWith("lsk");
		expect(walletData).toMatchSnapshot();
	});
});
