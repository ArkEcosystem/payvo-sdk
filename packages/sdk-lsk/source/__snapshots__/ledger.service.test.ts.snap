// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan should allow to pass a startPath 1`] = `
Object {
  "m/44'/134'/11'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "address": "lskr8nouvjpm5te9smbtj4satv43rwmhkdvydhuts",
      "balance": 0,
      "publicKey": "44e94b00d812b2c92c6e41d90e19aa9b849b81a3c719146fc657b497a6471705",
    },
  },
}
`;

exports[`scan should return scanned wallet 1`] = `
Object {
  "m/44'/134'/0'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "dpos": Object {
        "delegate": Object {
          "consecutiveMissedBlocks": 0,
          "isBanned": false,
          "lastForgedHeight": 14075260,
          "totalVotesReceived": "0",
          "username": "",
        },
      },
      "keys": Object {
        "mandatoryKeys": Array [],
        "numberOfSignatures": 0,
        "optionalKeys": Array [],
      },
      "sequence": Object {
        "nonce": "0",
      },
      "summary": Object {
        "address": "lsk8s6v2pdnxvab9oc42wbhvtb569jqg2ubjxgvvj",
        "balance": "6950000000",
        "isDelegate": false,
        "isMultisignature": false,
        "publicKey": null,
        "username": "",
      },
      "token": Object {
        "balance": "6950000000",
      },
    },
  },
  "m/44'/134'/1'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "dpos": Object {
        "delegate": Object {
          "consecutiveMissedBlocks": 0,
          "isBanned": false,
          "lastForgedHeight": 14075260,
          "totalVotesReceived": "0",
          "username": "",
        },
      },
      "keys": Object {
        "mandatoryKeys": Array [],
        "numberOfSignatures": 0,
        "optionalKeys": Array [],
      },
      "sequence": Object {
        "nonce": "0",
      },
      "summary": Object {
        "address": "lskbh47p4ts33c6c5pjvwa32424qr8px8pwfx8e4s",
        "balance": "1280000000",
        "isDelegate": false,
        "isMultisignature": false,
        "publicKey": null,
        "username": "",
      },
      "token": Object {
        "balance": "1280000000",
      },
    },
  },
  "m/44'/134'/2'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "dpos": Object {
        "delegate": Object {
          "consecutiveMissedBlocks": 0,
          "isBanned": false,
          "lastForgedHeight": 14075260,
          "totalVotesReceived": "0",
          "username": "",
        },
      },
      "keys": Object {
        "mandatoryKeys": Array [],
        "numberOfSignatures": 0,
        "optionalKeys": Array [],
      },
      "sequence": Object {
        "nonce": "0",
      },
      "summary": Object {
        "address": "lskksmfa2q2evtwmfneaon79u9hv7a3saokuy9tv9",
        "balance": "690000000",
        "isDelegate": false,
        "isMultisignature": false,
        "publicKey": null,
        "username": "",
      },
      "token": Object {
        "balance": "690000000",
      },
    },
  },
  "m/44'/134'/3'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "address": "lskrt9csby845j45crdat7j8mcnpgdk2w4yqhbwaj",
      "balance": 0,
      "publicKey": "5c12c704e97b79a7abd0cdb36336e4bdc675fcec5b8844d1fc95380c1d5db052",
    },
  },
}
`;

exports[`scan should support legacy 1`] = `
Object {
  "m/44'/134'/0'/0/0": WalletData {
    "bigNumberService": BigNumberService {
      "configRepository": ConfigRepository {},
    },
    "data": Object {
      "address": "7399986239080551550L",
      "balance": 0,
      "publicKey": "d48522677df50defd175c85072309c7643dbc6bdc63c7665a302579ed2ccaedb",
    },
  },
}
`;
